kroxylicious-schema-validation/src/main/java/io/kroxylicious/proxy/filter/schema/JsonValueValidator.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.filter.schema;
+
+import java.nio.ByteBuffer;
+
+import org.apache.kafka.common.record.Record;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.kroxylicious.proxy.filter.schema.validation.RecordValidator;
+
+/**
+ * Checks if a Record's value is valid JSON by attempting to deserialize it
+ * with jackson
+ */
+public class JsonValueValidator implements RecordValidator {
+
+    ObjectMapper mapper;
+
+    private JsonValueValidator() {
+        mapper = new ObjectMapper();
+    }
+
+    @Override
+    public Result validate(Record record) {
+        ByteBuffer value = record.value();
+        if (value == null) {
+            // maybe this should be configurable? a null value is often used to tombstone data but maybe users might
+            // want to prevent null values going to their topic
+            return VALID;
+        }
+        value.mark();
+        try {
+            byte[] bytes = new byte[record.valueSize()];
+            value.get(bytes);
+            mapper.readTree(bytes);
```

Comment:
Using `readTree` is more than you really need here. In particular it will do a bunch of allocation which can be avoided because you don't actually use the returned value. You can use the `JsonParser` API ([example](https://stackoverflow.com/questions/10226897/how-to-validate-json-with-jackson-json)) to check the JSON is well-formed but avoiding those allocations.

Reply:
I've reword to JsonParser, including some (dubiously useful) code to check if Object keys are unique. This appears to be a [recommendation](https://stackoverflow.com/questions/21832701/does-json-syntax-allow-duplicate-keys-in-an-object/23195243#23195243) in the JSON spec (but not a requirement) and I couldn't see an option to get JsonParser to do it for us (there is a `FAIL_ON_READING_DUP_TREE_KEY` deserialization feature, but it's only used when reading to Jackson TreeNode). I figured in some cases with dynamic JSON blobs you may still want to protect consumers from duplicate keys in objects as some parsers will choose to throw.